### PR TITLE
Create new buffer/definition model; Vesioned,Switchable,SoftDeletable; Source,Sink

### PR DESCRIPTION
**Pull request by michaljurecko**
Changes:
- Created new "definition" model to cover new requirements (high-throughput changes, SOX, ...) :
  - "Traits": 
    - **SoftDeletable**: [entity](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/softdeletable.go), [tests](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/softdeletable_test.go)
    - **Switchable**: [entity](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/switchable.go), [test](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/switchable_test.go)
    - **Versioned**: [entity](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/versioned.go), [test](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/versioned_test.go)
  - Entities:
    - **Branch**: entity, [repo](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/branch.go), [repo](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/repository/branch_repo.go), **[repo tests](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/repository/branch_repo_test.go)**
    - **Source**: entity [(1](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/source_base.go), [2](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/source_http.go)), [repo](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/repository/source_repo.go), **[repo tests](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/repository/source_repo_test.go)**
    - **Sink**: entity ([1](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/sink_base.go), [2](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/sink_table.go)), [repo](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/repository/sink_repo.go), **[repo tests](https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-286-definition-pkg/internal/pkg/service/buffer/definition/repository/sink_repo_test.go)**

---------------------------

Test coverage:

image

### CODE REVIEW

File: internal/pkg/service/common/etcdop/op/atomic.go [hunk 1]
```diff
@@ -87,6 +87,15 @@ func (v *AtomicOp) Write(factories ...func() Op) *AtomicOp {
...
 
+func (v *AtomicOp) BeforeWrite(fns ...func() error) *AtomicOp {
+	for _, fn := range fns {
+		v.WriteOrErr(func() (Op, error) {
+			return nil, fn()
+		})
+	}
+	return v
+}
+
```
Comment (michaljurecko) [inline internal/pkg/service/common/etcdop/op/atomic.go:98]: `Write`, alebo `WriteOrErr` moze vratit `op.Op=nil`, v tom pripade callback sluzi iba na nejaku kontrolu alebo pripravu dat, pre nasledujúce callback.

Aby to bolo v kode lepsie citatelne, tak som tu pridal tuto "skratku" `BeforeWrite`

File: internal/pkg/service/buffer/definition/key/branch.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+import (
+	"fmt"
+
+	"github.com/keboola/go-client/pkg/keboola"
+)
+
+type BranchKey struct {
+	ProjectID keboola.ProjectID `json:"projectId" validate:"required,min=1"`
+	BranchID  keboola.BranchID  `json:"branchId" validate:"required,min=1"`
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/key/branch.go:11]: Keys pre nove entity.

File: internal/pkg/service/buffer/definition/softdeletable.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+type SoftDeletableInterface interface {
+	IsDeleted() bool
+}
+
+type SoftDeletable struct {
+	Deleted bool `json:"deleted,omitempty"`
+	// DeletedWithParent is true when the object has not been deleted directly but was deleted together with its parent.
+	DeletedWithParent bool `json:"deletedWithParent,omitempty" validate:"excluded_if=Deleted false"`
+}
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/softdeletable.go:11]: Struktura, ktora bude `embedded` do `Branch,Source,Sink` entity, ktore maju podporovat soft-delete.

File: internal/pkg/service/buffer/definition/switchable.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	GetDisabledAt() *utctime.UTCTime
+	GetDisabledReason() string
+}
+
+type Switchable struct {
+	Disabled       bool             `json:"disabled,omitempty"`
+	DisabledBy     string           `json:"disabledBy,omitempty" required_if:"Disabled true"`
+	DisabledAt     *utctime.UTCTime `json:"disabledAt,omitempty"  required_if:"Disabled true"`
+	DisabledReason string           `json:"disabledReason,omitempty"  required_if:"Disabled true"`
+}
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/switchable.go:23]: `Source` a `Sink` ma byt mozne vypnut, ci uz v UI, alebo automaticky, ked je vyzadovana nejaka akcia uzivatela, napr. Storage API token bol refreshnuty a nevieme pokracovat v praci.

File: internal/pkg/service/buffer/definition/switchable.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+func (v *Switchable) Disable(now time.Time, by, reason string) {
+	at := utctime.From(now)
+	v.Disabled = true
+	v.DisabledBy = by
+	v.DisabledAt = &at
+	v.DisabledReason = reason
+}
+
+func (v *Switchable) GetDisabledBy() string {
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/switchable.go:44]: V Go sa nezvyknu getters prefixovat s `Get`, ale tu je konflikt s nazvom fieldu.

File: internal/pkg/service/buffer/definition/versioned.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+type VersionedInterface interface {
+	VersionNumber() VersionNumber
+	VersionHash() string
+	VersionModifiedAt() utctime.UTCTime
+	VersionDescription() string
+}
+
+type Versioned struct {
+	Version Version `json:"version"`
+}
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/versioned.go:21]: `Source` a `Sink` entity maju byt verziovane.

File: internal/pkg/service/buffer/definition/versioned.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+type Version struct {
+	Number      VersionNumber   `json:"number" hash:"ignore" validate:"required,min=1"`
+	Hash        string          `json:"hash" hash:"ignore" validate:"required,len=16"`
+	ModifiedAt  utctime.UTCTime `json:"modifiedAt" hash:"ignore" validate:"required"`
+	Description string          `json:"description" hash:"ignore"`
+}
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/versioned.go:28]: Samotna verzia nie je sucastou hashu, vsimni si `hash:"ignore"`, tj. ak je content rovnaky, ale cislo verzie je ine, tak hash bude rovnaky (testy su dalej).

File: internal/pkg/service/buffer/definition/branch.go [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package definition
+
+import (
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/definition/key"
+)
+
+type Branch struct {
+	key.BranchKey
+	SoftDeletable
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/branch.go:9]: `Branch` entity sluzi na to, aby som vedel povedat, ci bola branch uz inicializovana na pouzite so Stream API alebo nie.

Bez toho by som nevedel odlisit, neinicialitovanu branch od prazdnej.

Main branch sa vytvori (v dalsich PRs) automaticky pri prvom pouziti API.

File: internal/pkg/service/buffer/definition/source_base.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
+package definition
+
+import "github.com/keboola/keboola-as-code/internal/pkg/service/buffer/definition/key"
+
+type SourceType string
+
+type Source struct {
+	key.SourceKey
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/source_base.go:8]: `Source` entita, predtym sme to volali `Receiver`.

Je to pripravene, ze sa tam mozno neskor prida napr `KafkaSourceType`.

File: internal/pkg/service/buffer/definition/sink_base.go [hunk 10]
```diff
@@ -0,0 +1,18 @@
+package definition
+
+import (
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/definition/key"
+)
+
+type SinkType string
+
+type Sink struct {
+	key.SinkKey
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/sink_base.go:10]: `Sink` entita, predtym `Export`.

File: internal/pkg/service/buffer/definition/repository/repository.go [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+type dependencies interface {
+	Telemetry() telemetry.Telemetry
+	Clock() clock.Clock
+	EtcdClient() *etcd.Client
+	EtcdSerde() *serde.Serde
+}
+
+type Repository struct {
+	branch *BranchRepository
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/repository/repository.go:19]: Vytvorenie repository pkg, tento pkg je pokryty testami na `100%`.

File: internal/pkg/service/buffer/definition/repository/branch_repo.go [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+		telemetry: d.Telemetry(),
+		clock:     d.Clock(),
+		client:    d.EtcdClient(),
+		schema:    newBranchSchema(d.EtcdSerde()),
+		all:       all,
+	}
+}
+
+func (r *BranchRepository) List(parentKey keboola.ProjectID) iterator.DefinitionT[definition.Branch] {
+	return r.list(r.schema.Active(), parentKey)
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/repository/branch_repo.go:39]: V repositories sa pouziva vylepsena verzia `etcdop` frameworku, z predchadzajuceho [PR](https://github.com/keboola/keboola-as-code/pull/1441).

Vsetky operacie su atomicke (zalozene na etcd txn).

Comment (michaljurecko) [reply internal/pkg/service/buffer/definition/repository/branch_repo.go:39]: Kod repository aj testy vsetkych 3 entity: `Branch`,`Source`,`Sink` je prakticky rovnaky.

Dala by sa z toho spravit sablona a repository aj testy generovat cez `go generate`, ale kedze mam iba 3 entity, a nevyzera to, ze ich bude casom nejak viac, tak to nebudem teraz prehanat.

File: internal/pkg/service/buffer/definition/repository/branch_repo_test.go [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+	branch2.BranchKey = key.BranchKey{
+		ProjectID: 456,
+		BranchID:  789,
+	}
+
+	clk := clock.NewMock()
+	clk.Set(utctime.MustParse("2006-01-02T15:04:05.123Z").Time())
+
+	d := deps.NewMocked(t, deps.WithEnabledEtcdClient(), deps.WithClock(clk))
+	client := d.TestEtcdClient()
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/repository/branch_repo_test.go:48]: Testy su nieco medzi unit a E2E.
Zapisuje sa do realnej etcd database (test ma credentials cez ENV, vid `etcdhelper` pkg), no ostatne dependencies, napr. `clock` su mockovane.

Kazdy test ma svoj `etcd namespace`, ktory sa po teste zmaze.

File: internal/pkg/service/buffer/definition/repository/branch_repo_test.go [hunk 14]
```diff
@@ -0,0 +1,431 @@
...
+		assert.Empty(t, branches)
+	}
+	{
+		// Get - not found
+		if err := r.Get(branch1.BranchKey).Do(ctx).Err(); assert.Error(t, err) {
+			assert.Equal(t, `branch "123/567" not found in the project`, err.Error())
+			var errWithStatus serviceErrors.WithStatusCode
+			if assert.True(t, errors.As(err, &errWithStatus)) {
+				assert.Equal(t, http.StatusNotFound, errWithStatus.StatusCode())
+			}
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/repository/branch_repo_test.go:78]: Logicke chyby maju nastaveny HTTP kod.
Takze ich v API vrstve mozem priamo vratit, bez potreby dalsieho spracovania.

File: internal/pkg/service/buffer/definition/repository/source_repo.go [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+func (r *SourceRepository) Version(k key.SourceKey, version definition.VersionNumber) op.ForType[*op.KeyValueT[definition.Source]] {
+	return r.schema.
+		Versions().Of(k).Version(version).Get(r.client).
+		WithEmptyResultAsError(func() error {
+			return serviceError.NewResourceNotFoundError("source version", k.String()+"/"+version.String(), "branch")
+		})
+}
+
+func (r *SourceRepository) Rollback(k key.SourceKey, to definition.VersionNumber) *op.AtomicOp {
+	var v definition.Source
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/repository/source_repo.go:257]: `Source` a `Sink` entity maju oproti `Branch` naviac aj versioning.

File: internal/pkg/service/buffer/definition/repository/source_repo_test.go [hunk 16]
```diff
@@ -0,0 +1,634 @@
...
+>>>>>
+
+<<<<<
+definition/branch/active/123/789
+-----
+%A
+>>>>>
+
+<<<<<
+definition/source/active/123/456/my-source-1
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/definition/repository/source_repo_test.go:444]: Schema etcd databazy je zvolena tak, ze jednou `watch` operaciou mozem sledovat vsetky aktivne objekty.
Zmazane objekty a všetky verzie su ulozene bokom.

File: internal/pkg/service/buffer/storage/statistics/cache/cache_test.go [hunk 17]
```diff
@@ -54,7 +54,7 @@ func TestCaches(t *testing.T) {
 		{
 			Description: "Empty",
 			Assert: func(provider repository.Provider) {
-				stats, err := provider.ExportStats(ctx, sliceKey1.ExportKey)
+				stats, err := provider.SinkStats(ctx, sliceKey1.SinkKey)
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/statistics/cache/cache_test.go:57]: V tomto commite je upraveny `storage/statistics` pkg.
- premenovane receiver,export -> source,sink
- statistiky su uz tiez branchovane

File: internal/pkg/service/buffer/definition/softdeletable.go (hunk 3 above)
Comment (michaljurecko) [reply internal/pkg/service/buffer/definition/softdeletable.go:11]: `DeletedWithParent bool`

Ked sa zmaze parent object, napr. `branch`, ta sa musia zmazat aj vsetky pod objekty.
`branch -> source -> sink`.

Ked spravim `undelete branch`, tak sa ma vsetko obnovit.

No nemozem obnovit tie `sources/sinks`, ktore boli zmazane uz skor a zamerne (teda nie vramci mazania parent objektu).

Na rozlisenie tychto 2 pripadov sluzi `DeletedWithParent` flag.

-------------------

Skusil som pozriet ako to riesia a aky naming tomu davaju inde, [tu](https://laravel-news.com/laravel-soft-delete-parent) je to napr. `parent_deleted_at`.

**Review by zajca — APPROVED** [review]

**Pull request merged**